### PR TITLE
Replace MAX_BYTE_SIZE constant with setting

### PR DESCRIPTION
**Pull request by gitlab-djensen**
## Status
**READY**

## Migrations
NO

## Description
The MAX_BYTE_SIZE constant did not allow for customization, which
is necessary for cases where legitimate SAML responses are larger
than 250,000 bytes. This replaces the constant with a setting, which
has a default value of 250,000 bytes, but can be customized like
any other setting.

This was motivated by https://github.com/onelogin/ruby-saml/issues/598, which was motivated by https://gitlab.com/gitlab-org/gitlab/-/issues/329053.
This follows the [existing pattern used for `encode_raw_saml`](https://github.com/onelogin/ruby-saml/blob/master/lib/onelogin/ruby-saml/saml_message.rb#L111).

## Related PRs
List related PRs against other branches:

branch | PR
------ | ------
Introduce MAX_BYTE_SIZE | [link](https://github.com/onelogin/ruby-saml/pull/566/files#diff-6e890879e7944b6cac0e115f001b790f412dc5f6a9a4a11651f70f31c6a33124R25)
Introduce inflate option to disable SAML message inflation | [link](https://github.com/onelogin/ruby-saml/pull/383)

## Todos
- [x] Tests
- [x] Documentation

## Deploy Notes
Nothing noteworthy.

## Steps to Test or Reproduce
Outline the steps to test or reproduce the PR here.

```sh
git fetch
git checkout gitlab-djensen:gitlab-djensen-use-setting-for-max-byte-size
bundle install;
ruby -Itest test/settings_test.rb
```

## Impacted Areas in Application
List general components of the application that this PR will affect:

* Settings
* SamlMessage

### CODE REVIEW

**Comment by gitlab-djensen** [comment]
@dblessing would you mind doing the initial review on this whenever you have some time?

**Comment by dblessing** [comment]
@gitlab-djensen Thanks for opening the PR. Looks good to me, with my limited knowledge of the `ruby-saml` code. Let's see what the maintainers think.

**Comment by gitlab-djensen** [comment]
@pitbulk this PR is ready for your review, whenever you might have some time.

GitLab is hoping you will release a new version of this gem containing this fix, because `MAX_BYTE_SIZE` was a breaking change for implementations with large messages ([example](https://gitlab.com/gitlab-org/gitlab/-/issues/329053)).

**Comment by johnnyshields** [comment]
👍 looks good to merge

**Comment by gitlab-djensen** [comment]
@pitbulk friendly ping, hopefully this is an easy review 🙂

File: lib/onelogin/ruby-saml/saml_message.rb
Comment (pitbulk) [inline lib/onelogin/ruby-saml/saml_message.rb:91]: instead the whole setting object, can we just add message_max_bytesize param by default MAX_BYTE_SIZE?

Comment (gitlab-djensen) [reply lib/onelogin/ruby-saml/saml_message.rb:91]: @pitbulk I'm pretty sure we could, but this follows the [existing pattern used for `encode_raw_saml`](https://github.com/onelogin/ruby-saml/blob/master/lib/onelogin/ruby-saml/saml_message.rb#L111) (the inverse method).  Do you want me to change this method even though it would then differ from `encode_raw_saml`?

Comment (pitbulk) [reply lib/onelogin/ruby-saml/saml_message.rb:91]: ohh you are right, let's then keep it as it is

File: lib/onelogin/ruby-saml/settings.rb
Comment (pitbulk) [inline lib/onelogin/ruby-saml/settings.rb:262]: Can this default value be read from SamlMessage.MAX_BYTE_SIZE?

Comment (gitlab-djensen) [reply lib/onelogin/ruby-saml/settings.rb:262]: @pitbulk ha, I missed that! Just pushed a commit that removes it 👍

**Comment by gitlab-djensen** [comment]
@pitbulk thanks for reviewing!  I just marked both of your comments as resolved.  Should we run the workflow next?  (It's telling me "First-time contributors need a maintainer to approve running workflows.")

**Comment by gitlab-djensen** [comment]
Strangely, [the workflow ran anyway](https://github.com/onelogin/ruby-saml/actions/runs/1125004867).  There were a bunch of failures because `settings` was still optional in some places where it is now required.  I just pushed 2 commits that make `settings` a required argument in multiple places, **which makes this PR a breaking change**.   (We could avoid the breaking change by leaving `settings` as an optional argument and calling `OneLogin::RubySaml::Settings.new` when `settings` is not provided.  But that approach is not currently used anywhere in the gem, so it would be a new pattern.)  I'll keep working on this until the tests are green.

**Comment by pitbulk** [comment]
We need to avoid breaking changes. If settings is not provided to the decode_raw_saml method, makes sense to keep using the default value, otherwise, the settings object should be there.

**Comment by gitlab-djensen** [comment]
@pitbulk gotcha, refactored to be non-breaking (by calling `OneLogin::RubySaml::Settings.new` when `settings` is not passed).  Tests are green.  How's it looking now?

**Comment by johnnyshields** [comment]
@pitbulk this looks good to me.

**Comment by gitlab-djensen** [comment]
@pitbulk friendly ping, [prompted by a colleague](https://gitlab.com/gitlab-org/gitlab/-/issues/329053#note_667767935) with an affected client.  Does this look good for merge?

**Comment by johnnyshields** [comment]
Also, after merge, let's get an gem release please :) Lots of great improvements recently.

**Comment by pitbulk** [comment]
Sorry I will be back from PTO on 6 and take care of all the pending PRs

**Review by pitbulk — APPROVED** [review]

**Pull request merged**
